File: Python/316.remove-duplicate-letters.py
#
# @lc app=leetcode id=316 lang=python
#
# [316] Remove Duplicate Letters
#
# https://leetcode.com/problems/remove-duplicate-letters/description/
#
# algorithms
# Hard (33.69%)
# Likes:    933
# Dislikes: 87
# Total Accepted:    65K
# Total Submissions: 192.9K
# Testcase Example:  '"bcabc"'
#
# Given a string which contains only lowercase letters, remove duplicate
# letters so that every letter appears once and only once. You must make sure
# your result is the smallest in lexicographical order among all possible
# results.
# 
# Example 1:
# 
# 
# Input: "bcabc"
# Output: "abc"
# 
# 
# Example 2:
# 
# 
# Input: "cbacdcbc"
# Output: "acdb"
# 
# 
#

# @lc code=start
from heapq import *
class Solution(object):
    def removeDuplicateLetters(self, s):
        """
        :type s: str
        :rtype: str
        """
        # min_heap = []
        # for c in s:
        #     # item = (ord(c)-97, c)
        #     if c not in min_heap:
        #         heappush(min_heap, c)
            
        #         # heappush(min_heap, (matrix[i][0], 0, matrix[i]))
        # res = ''
        # while min_heap:
        #     res += heappop(min_heap)
        # return res 

        # keep the relative positiion 
        pos = {c: i for i, c in enumerate(s)}
        stack = []
        for i, c in enumerate(s):
            if c not in stack:
                while stack and c < stack[-1] and i < pos[stack[-1]]:
                    stack.pop()
                stack.append(c)
        return ''.join(stack)

        # 厉害啊 [-1] -> [-1:]
        # rindex = {c: i for i, c in enumerate(s)}
        # result = ''
        # for i, c in enumerate(s):
        #     if c not in result:
        #         while c < result[-1:] and i < rindex[result[-1]]:
        #             result = result[:-1]
        #         result += c
        # return result

# @lc code=end

sol = Solution()
s = 'cbacdcbc'
print(sol.removeDuplicateLetters(s))

File: C++/316.remove-duplicate-letters.cpp
/*
 * @lc app=leetcode id=316 lang=cpp
 *
 * [316] Remove Duplicate Letters
 *
 * https://leetcode.com/problems/remove-duplicate-letters/description/
 *
 * algorithms
 * Hard (35.95%)
 * Likes:    1501
 * Dislikes: 115
 * Total Accepted:    83.6K
 * Total Submissions: 231K
 * Testcase Example:  '"bcabc"'
 *
 * Given a string which contains only lowercase letters, remove duplicate
 * letters so that every letter appears once and only once. You must make sure
 * your result is the smallest in lexicographical order among all possible
 * results.
 * 
 * Example 1:
 * 
 * 
 * Input: "bcabc"
 * Output: "abc"
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: "cbacdcbc"
 * Output: "acdb"
 * 
 * 
 * Note: This question is the same as 1081:
 * https://leetcode.com/problems/smallest-subsequence-of-distinct-characters/
 * 
 */

// @lc code=start
#include<vector>
#include<string>
#include<unordered_map>
using namespace std;

// @lc code=start
class Solution {
public:
    string removeDuplicateLetters(string s) {
        if (s.empty()) return s;
        
        unordered_map<char, int> cnts;
        for (char ch: s) cnts[ch] ++;
        bool visited[256] = {false};
        
        string ans{s[0]};
        visited[s[0]] = true;
        cnts[s[0]] --;
        
        for (int i=1; i<s.size(); ++i) {
            char ch = s[i];
            cnts[ch] --;
            if (visited[ch]) continue;
            
            while (!ans.empty() && ch<ans.back() && cnts[ans.back()]>0) {
                visited[ans.back()] = false;
                ans.pop_back();
            }
            ans.push_back(ch);
            visited[ch] = true;
        }
        return ans;
    }
};
// @lc code=end

